File: frontend/src/components/calendar/holidays.ts
import type { Holiday } from '../../types';

export function computeEaster(year: number): Date {
  const a = year % 19, b = Math.floor(year / 100), c = year % 100;
  const d = Math.floor(b / 4), e = b % 4, f = Math.floor((b + 8) / 25);
  const g = Math.floor((b - f + 1) / 3), h = (19 * a + b - d - g + 15) % 30;
  const i = Math.floor(c / 4), k = c % 4;
  const l = (32 + 2 * e + 2 * i - h - k) % 7;
  const m = Math.floor((a + 11 * h + 22 * l) / 451);
  const month = Math.floor((h + l - 7 * m + 114) / 31);
  const day = ((h + l - 7 * m + 114) % 31) + 1;
  return new Date(year, month - 1, day);
}

export function getHolidaysForYear(year: number): Holiday[] {
  const easter = computeEaster(year);
  const ed = (offset: number) => {
    const d = new Date(easter);
    d.setDate(d.getDate() + offset);
    return d.toISOString().slice(0, 10);
  };
  const f = (m: number, d: number) =>
    `${year}-${String(m).padStart(2, '0')}-${String(d).padStart(2, '0')}`;

  return [
    { date: f(1, 1), name: 'Neujahr', nameEN: "New Year's Day" },
    { date: ed(-2), name: 'Karfreitag', nameEN: 'Good Friday' },
    { date: ed(1), name: 'Ostermontag', nameEN: 'Easter Monday' },
    { date: f(5, 1), name: 'Tag der Arbeit', nameEN: 'Labour Day' },
    { date: ed(39), name: 'Christi Himmelfahrt', nameEN: 'Ascension Day' },
    { date: ed(50), name: 'Pfingstmontag', nameEN: 'Whit Monday' },
    { date: f(10, 3), name: 'Tag der Deutschen Einheit', nameEN: 'German Unity Day' },
    { date: f(10, 31), name: 'Reformationstag', nameEN: 'Reformation Day' },
    { date: f(12, 25), name: '1. Weihnachtstag', nameEN: 'Christmas Day' },
    { date: f(12, 26), name: '2. Weihnachtstag', nameEN: "St. Stephen's Day" },
  ];
}

export function getDaysInMonth(y: number, m: number) {
  return new Date(y, m + 1, 0).getDate();
}

export function getFirstDayOfWeek(y: number, m: number) {
  const d = new Date(y, m, 1).getDay();
  return d === 0 ? 6 : d - 1;
}

export function formatDateStr(y: number, m: number, d: number) {
  return `${y}-${String(m + 1).padStart(2, '0')}-${String(d).padStart(2, '0')}`;
}

export const MONTHS = ['January', 'February', 'March', 'April', 'May', 'June', 'July', 'August', 'September', 'October', 'November', 'December'];
export const DAYS = ['Mon', 'Tue', 'Wed', 'Thu', 'Fri', 'Sat', 'Sun'];
export const EVENT_COLORS = ['#22d3ee', '#a78bfa', '#fb7185', '#4ade80', '#fbbf24', '#60a5fa', '#f97316'];

File: internal/holidays/holidays.go
package holidays

import (
	"fmt"
	"time"
)

// Holiday represents a public holiday.
type Holiday struct {
	Date   string `json:"date"`
	Name   string `json:"name"`
	NameEN string `json:"nameEN"`
}

// easterSunday calculates the date of Easter Sunday for a given year
// using the anonymous Gregorian algorithm (Computus).
func easterSunday(year int) time.Time {
	a := year % 19
	b := year / 100
	c := year % 100
	d := b / 4
	e := b % 4
	f := (b + 8) / 25
	g := (b - f + 1) / 3
	h := (19*a + b - d - g + 15) % 30
	i := c / 4
	k := c % 4
	l := (32 + 2*e + 2*i - h - k) % 7
	m := (a + 11*h + 22*l) / 451
	month := (h + l - 7*m + 114) / 31
	day := ((h + l - 7*m + 114) % 31) + 1
	return time.Date(year, time.Month(month), day, 0, 0, 0, 0, time.UTC)
}

func formatDate(t time.Time) string {
	return fmt.Sprintf("%04d-%02d-%02d", t.Year(), t.Month(), t.Day())
}

func addDays(t time.Time, days int) time.Time {
	return t.AddDate(0, 0, days)
}

// GetHolidays returns all public holidays for Niedersachsen for the given year.
// Covers fixed holidays and Easter-based movable holidays.
func GetHolidays(year int) []Holiday {
	easter := easterSunday(year)

	holidays := []Holiday{
		// Fixed holidays
		{
			Date:   fmt.Sprintf("%04d-01-01", year),
			Name:   "Neujahr",
			NameEN: "New Year's Day",
		},
		{
			Date:   fmt.Sprintf("%04d-05-01", year),
			Name:   "Tag der Arbeit",
			NameEN: "Labour Day",
		},
		{
			Date:   fmt.Sprintf("%04d-10-03", year),
			Name:   "Tag der Deutschen Einheit",
			NameEN: "German Unity Day",
		},
		{
			Date:   fmt.Sprintf("%04d-10-31", year),
			Name:   "Reformationstag",
			NameEN: "Reformation Day",
		},
		{
			Date:   fmt.Sprintf("%04d-12-25", year),
			Name:   "1. Weihnachtstag",
			NameEN: "Christmas Day",
		},
		{
			Date:   fmt.Sprintf("%04d-12-26", year),
			Name:   "2. Weihnachtstag",
			NameEN: "Boxing Day",
		},
		// Easter-based movable holidays
		{
			Date:   formatDate(addDays(easter, -2)),
			Name:   "Karfreitag",
			NameEN: "Good Friday",
		},
		{
			Date:   formatDate(addDays(easter, 1)),
			Name:   "Ostermontag",
			NameEN: "Easter Monday",
		},
		{
			Date:   formatDate(addDays(easter, 39)),
			Name:   "Christi Himmelfahrt",
			NameEN: "Ascension Day",
		},
		{
			Date:   formatDate(addDays(easter, 50)),
			Name:   "Pfingstmontag",
			NameEN: "Whit Monday",
		},
	}

	return holidays
}

// GetHolidaysRange returns all public holidays for Niedersachsen
// for each year in the inclusive range [startYear, endYear].
func GetHolidaysRange(startYear, endYear int) []Holiday {
	var result []Holiday
	for year := startYear; year <= endYear; year++ {
		result = append(result, GetHolidays(year)...)
	}
	return result
}
